akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala

```diff
@@ -149,6 +133,90 @@ private[http] object OutgoingConnectionBlueprint {
 
   import ParserOutput._
 
+  private final class ResponsePrep(parserSettings: ParserSettings, eagerClose: Boolean)
+    extends GraphStage[FlowShape[ResponseOutput, HttpResponse]] {
+
+    private val in = Inlet[ResponseOutput]("ResponsePrep.in")
+    private val out = Outlet[HttpResponse]("ResponsePrep.out")
+
+    val shape = new FlowShape(in, out)
+
+    override def createLogic(effectiveAttributes: Attributes) = new GraphStageLogic(shape) {
+      private val idle = new InHandler {
+        def onPush(): Unit = grab(in) match {
+          case ResponseStart(statusCode, protocol, headers, entityCreator, _) ⇒
+            val entity = createEntity(entityCreator) withSizeLimit parserSettings.maxContentLength
+            push(out, HttpResponse(statusCode, headers, entity, protocol))
+
+          case MessageStartError(_, info) ⇒
+            throw IllegalResponseException(info)
+        }
+      }
+
+      private val waitForMessageEnd = new InHandler {
+        def onPush(): Unit = grab(in) match {
+          case MessageEnd ⇒ setHandler(in, idle)
+        }
+      }
+
+      setHandler(in, idle)
+
+      private def createEntity(creator: EntityCreator[ResponseOutput, ResponseEntity]): ResponseEntity = {
+        creator match {
+          case StrictEntityCreator(entity) ⇒
+            pull(in)
+            setHandler(in, waitForMessageEnd)
+            entity
+
+          case StreamedEntityCreator(creator) ⇒
+            val entitySource = new SubSourceOutlet[ResponseOutput]("EntitySource")
+            var responseCancelled = false
+
+            entitySource.setHandler(new OutHandler {
+              override def onPull(): Unit = pull(in)
+
+              override def onDownstreamFinish(): Unit = {
+                responseCancelled = true
+                if (isClosed(in)) completeStage()
+                else {
+                  // Start draining
+                  if (!hasBeenPulled(in)) pull(in)
+                }
+              }
+            })
+
+            setHandler(in, new InHandler {
+              override def onPush(): Unit = grab(in) match {
+                case MessageEnd ⇒
+                  entitySource.complete()
+                  setHandler(in, idle)
+                  responseCancelled = false
+
+                case messagePart ⇒
+                  if (responseCancelled) {
+                    if (eagerClose) completeStage()
```

Comment:
all completions must be accompanied by closing the entitySource as well, else we leak it